core/Azure.Mcp.Core/src/Areas/Server/Commands/ServiceStartCommand.cs

```diff
@@ -326,48 +373,163 @@ private IHost CreateStdioHost(ServiceStartOptions serverOptions)
-                    app.UseCors("AllowAll");
-                    app.UseRouting();
-                    app.UseEndpoints(endpoints =>
+                    IOptionsMonitor<MicrosoftIdentityOptions> azureAdOptionsMonitor = context.RequestServices.GetRequiredService<IOptionsMonitor<MicrosoftIdentityOptions>>();
+                    MicrosoftIdentityOptions azureAdOptions = azureAdOptionsMonitor.Get(JwtBearerDefaults.AuthenticationScheme);
+                    HttpRequest request = context.Request;
+                    string baseUrl = $"{request.Scheme}://{request.Host}";
+                    string? clientId = azureAdOptions.ClientId;
+                    string? tenantId = azureAdOptions.TenantId;
+                    string instance = azureAdOptions.Instance?.TrimEnd('/') ?? "https://login.microsoftonline.com";
+
+                    var metadata = new OAuthProtectedResourceMetadata
                     {
-                        endpoints.MapMcp();
-                    });
-                });
+                        Resource = baseUrl,
+                        AuthorizationServers = [$"{instance}/{tenantId}/v2.0"],
+                        ScopesSupported = [$"{clientId}/MCP.All"],
+                        BearerMethodsSupported = ["header"],
+
+                        // Intentionally pointing to MCP repo for documentation. Could eventually
+                        // have a dedicated usage doc page, potentially provided by this service itself.
+                        ResourceDocumentation = "https://github.com/Microsoft/mcp",
+                        ResourceSigningAlgValuesSupported = ["RS256"]
+                    };
+
+                    context.Response.ContentType = "application/json";
+                    await JsonSerializer.SerializeAsync(
+                        context.Response.Body,
+                        metadata,
+                        OAuthMetadataJsonContext.Default.OAuthProtectedResourceMetadata);
+                    return;
+                }
 
-                var url = GetSafeAspNetCoreUrl();
-                webBuilder.UseUrls(url);
-            })
-            .Build();
-    }
+                await next(context);
+            });
+        }
+
+        // AuthN/Z are always required in the remote HTTP service scenario.
+        if (serverOptions.RunAsRemoteHttpService)
+        {
+            app.UseAuthentication();
+            app.UseAuthorization();
+        }
+
+        IEndpointConventionBuilder mcpEndpointBuilder = app.MapMcp();
+        if (serverOptions.RunAsRemoteHttpService)
+        {
+            // All MCP endpoints require MCP.All scope or role
+            mcpEndpointBuilder.RequireAuthorization("McpAccess");
+        }
+
+        // Map non-MCP endpoints as needed.
+        if (serverOptions.RunAsRemoteHttpService)
+        {
+            // Health checks are anonymous (no authentication required)
+            app.MapHealthChecks("/health")
+                .AllowAnonymous();
+        }
```

Comment:
nit: wondering if it makes sense to merge these two if statements.

```C#
if (serverOptions.RunAsRemoteHttpService)
{
   // All MCP endpoints require MCP.All scope or role
   app.MapMcp().RequireAuthorization("McpAccess"); 
   // Health checks are anonymous (no authentication required)
   app.MapHealthChecks("/health").AllowAnonymous();

}
```

Reply:
I was considering that, but I also wanted to make it very clear for the uninitiated the pieces that are "for MCP goodness" vs "for regular web API goodness" at this part in the middleware setup. I was using conditional blocks similar to how some folks use `#region`.

Reply:
I see - thanks for the explaining, that makes sense.